src/linker/Linker/TypeReferenceExtensions.cs

```diff
@@ -361,6 +361,11 @@ public static bool IsTypeOf<T> (this TypeReference tr)
 
 		public static bool IsTypeOf (this TypeReference tr, WellKnownType type)
 		{
+			// TypeReferences of System.Array doesn't have metadata type of MetadataType.Array
+			if (type.TryGetMetadataType (out var metadataType)) {
+				Debug.Assert (tr.IsTypeOf (type.GetNamespace (), type.GetName ()) == (metadataType == tr.MetadataType));
+				return metadataType == tr.MetadataType;
+			}
 			var (@namespace, name) = type.GetNamespaceAndName ();
 			return tr.IsTypeOf (@namespace, name);
```

Comment:
Can this be a simple switch instead of this please?

Reply:
I'm not sure I understand what part here can be turned into a switch. Can you explain what you mean?